### PR TITLE
feat: lazy initialize vector builder on write

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Lazy create `MutableVector` for `TimeSeries`. This change can save lots of memory when the data's field columns are sparse. In my test scenario where ~90% fields are null, it can cut up to 80% memory consumption:
 
Snipaste_2024-01-21_15-20-39

I also run a tsbs-load benchmark to see if this change slow down the normal case. The result shows it doesn't have bad influence:

before this change:
```
loaded 86400000 rows in 141.686sec with 6 workers (mean rate 609798.27 rows/sec)
```
after this change:
```
loaded 86400000 rows in 140.448sec with 6 workers (mean rate 615174.51 rows/sec)
```

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/memtable/time_series.rs [hunk 1]
```diff
@@ -620,16 +622,19 @@ impl ValueBuilder {
...
         let op_type = UInt8VectorBuilder::with_capacity(capacity);
 
-        let fields = region_metadata
+        let field_types = region_metadata
             .field_columns()
-            .map(|c| c.column_schema.data_type.create_mutable_vector(capacity))
-            .collect();
+            .map(|c| c.column_schema.data_type.clone())
+            .collect::<Vec<_>>();
+        let fields = (0..field_types.len()).map(|_| None).collect();
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:629]:
```suggestion
        let fields = vec![None; field_types.len()];
```

File: src/metric-engine/src/engine/create.rs [hunk 2]
```diff
@@ -363,6 +363,7 @@ impl MetricEngineInner {
             .for_each(|metadata| {
                 if metadata.semantic_type == SemanticType::Tag {
                     metadata.semantic_type = SemanticType::Field;
+                    metadata.column_schema.is_nullable();
```
Comment (v0y4g3r) [inline src/metric-engine/src/engine/create.rs:366]:
```suggestion
```

File: src/mito2/src/memtable/time_series.rs [hunk 3]
```diff
@@ -640,8 +645,26 @@ impl ValueBuilder {
...
+                    .get_or_insert_with(|| {
+                        // lazy initialize on first non-null value
+                        let mut mutable_vector = self
+                            .region_metadata
+                            .field_columns()
+                            .nth(idx)
+                            .unwrap()
+                            .column_schema
+                            .data_type
+                            .create_mutable_vector(num_rows);
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:661]: Can we use `field_types`?

File: src/mito2/src/memtable/time_series.rs (hunk 1 above)
Comment (waynexia) [reply src/mito2/src/memtable/time_series.rs:629]: `MutableVector` is not copy-able

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**